### PR TITLE
Account for inverse rate

### PR DESCRIPTION
**Pull request by JavedNissar**
This is a PR to address #51. Feel free to close.

### CODE REVIEW

File: src/exchange.rs [hunk 1]
```diff
@@ -2,11 +2,46 @@ use crate::currency::FormattableCurrency;
...
+    to: &'a T,
+}
+
+impl<'a, T: FormattableCurrency> PartialEq for ExchangeRateQuery<'a, T>{
+    fn eq(&self, other: &Self) -> bool {
+        self.from == other.from && self.to == other.to
+    }
+}
+
+impl<'a, T: FormattableCurrency> Eq for ExchangeRateQuery<'a, T>{}
```
Comment (varunsrin) [inline src/exchange.rs:20]: could you use the derive macro here if it's equivalent?  

```
#[derive(Eq)]
struct ExchangeRateQuery<'a, T: FormattableCurrency>{
    from: &'a T,
    to: &'a T,
}
```

File: src/exchange.rs [hunk 2]
```diff
@@ -2,11 +2,46 @@ use crate::currency::FormattableCurrency;
 use crate::{Money, MoneyError};
 use rust_decimal::Decimal;
 use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+use std::fmt;
+use rust_decimal_macros::*;
+
+struct ExchangeRateQuery<'a, T: FormattableCurrency>{
```
Comment (varunsrin) [inline src/exchange.rs:9]: What do you feel is the primary benefit of using a struct vs just passing in the currencies and letting Exchange handle the complexity?

File: src/exchange.rs [hunk 3]
```diff
@@ -18,22 +53,37 @@ impl<'a, T: FormattableCurrency> Exchange<'a, T> {
 
     /// Update an ExchangeRate or add it if does not exist.
     pub fn set_rate(&mut self, rate: &'a ExchangeRate<T>) {
-        let key = Exchange::generate_key(rate.from, rate.to);
-        self.map.insert(key, *rate);
+        let query = ExchangeRateQuery::new(rate.from, rate.to);
```
Comment (varunsrin) [inline src/exchange.rs:56]: the default behavior of always setting the inverse rate might not be desirable. for instance, if there is an exchange rate spread then the inverse rate is not always the same as the forward rate. 

do you think there's a reasonable interface where we can make setting the exact inverse an optional parameter?

**Review by varunsrin — CHANGES_REQUESTED** [review]
First off, thanks for submitting this, appreciate the PR! 

Meta thoughts on this: 

1. What is the primary benefit of doing the ExchangeRateQuery over just building the logic into the Exchange in your opinion? I'm trying to justify the added complexity of another class here.  

2. On further reflection, forcing the inverse rate to always be set isn't desirable for all cases (e.g. currency spreads). I think this should be an optional flag on set_rate. 

3. Since this is a big API change, ill hold this back till the 0.5.0 release if we do decide to go forward with it.

File: src/exchange.rs (hunk 2 above)
Comment (RestitutorOrbis) [reply src/exchange.rs:9]: Admittedly, I mostly did this is to play around with implementing keys for HashMap since I'm still new to Rust; however, I do think using a dedicated type is more idiomatic and feels more ergonomic. That being said, if you want to switch over to using string keys that's fine by me.

File: src/exchange.rs (hunk 3 above)
Comment (RestitutorOrbis) [reply src/exchange.rs:56]: That sounds reasonable to me. I'm envisioning two options for this:
```
pub fn set_rate(&mut self, rate: &'a ExchangeRate<T>, set_inverse: bool)
```
which forces the user to explicitly pass in whether or not they want the inverse which feels a bit un-ergonomic to me and the other option is to add a new function with the following signature:
```
pub fn set_rate_and_inverse_rate(&mut self, rate: &'a ExchangeRate<T>)
```
which would both set the rate as well as set the inverse rate.

**Comment by RestitutorOrbis** [comment]
> First off, thanks for submitting this, appreciate the PR!
> 
> Meta thoughts on this:
> 
> 1. What is the primary benefit of doing the ExchangeRateQuery over just building the logic into the Exchange in your opinion? I'm trying to justify the added complexity of another class here.
> 2. On further reflection, forcing the inverse rate to always be set isn't desirable for all cases (e.g. currency spreads). I think this should be an optional flag on set_rate.
> 3. Since this is a big API change, ill hold this back till the 0.5.0 release if we do decide to go forward with it.

You're welcome! Thanks for publishing this library, it's great! With regard to those meta thoughts:
1. I think I covered this in a reply to one of your line comments
2. Also in a line comment reply
3. That's totally fine!

File: src/exchange.rs (hunk 3 above)
Comment (varunsrin) [reply src/exchange.rs:56]: let's go with the latter, it also avoids a breaking api change

File: src/exchange.rs (hunk 2 above)
Comment (varunsrin) [reply src/exchange.rs:9]: If we can stick with using the strings and the same interface and avoiding a new type, we can merge the change safely into the 0.4.2 release without worrying about backward compat fixes. 

i'm rethinking the exchange design for the next big release (0.5), so perhaps we can revisit this idea then.

File: src/exchange.rs [hunk 4]
```diff
@@ -118,6 +168,12 @@ mod tests {
 
         let fetched_rate = exchange.get_rate(usd, gbp).unwrap();
         assert_eq!(fetched_rate.rate, dec!(1.6));
+
+        let fetched_reverse_rate = exchange.get_rate(eur, usd).unwrap();
+        assert_eq!(fetched_reverse_rate.rate, dec!(1) / dec!(1.5));
+
+        let fetched_reverse_rate = exchange.get_rate(gbp, usd).unwrap();
+        assert_eq!(fetched_reverse_rate.rate, dec!(1) / dec!(1.6));
```
Comment (varunsrin) [inline src/exchange.rs:176]: i dont know if we handle the case where someone sets the exchange rate to 0 well. it was probably OK when we didn't have an inverse rate but since this will raise new errors (1/0) we should add explicit tests cases for that.

File: src/exchange.rs [hunk 5]
```diff
@@ -22,6 +23,20 @@ impl<'a, T: FormattableCurrency> Exchange<'a, T> {
         self.map.insert(key, *rate);
     }
 
+    pub fn set_rate_and_inverse(&mut self, rate: &'a ExchangeRate<T>) -> Result<ExchangeRate<'a, T>, MoneyError> {
+        if rate.rate == dec!(0) {
+            return Err(MoneyError::DivideByZero)
```
Comment (RestitutorOrbis) [inline src/exchange.rs:28]: How would you like to see divide by zero handled? I felt like returning a `MoneyError` made the most sense but I'm curious to hear your thoughts.

File: src/exchange.rs [hunk 6]
```diff
@@ -22,6 +23,20 @@ impl<'a, T: FormattableCurrency> Exchange<'a, T> {
         self.map.insert(key, *rate);
     }
 
+    pub fn set_rate_and_inverse(&mut self, rate: &'a ExchangeRate<T>) -> Result<ExchangeRate<'a, T>, MoneyError> {
```
Comment (RestitutorOrbis) [inline src/exchange.rs:26]: Return a result with inverse rate or `MoneyError` since it would be weird to return `Option<MoneyError>` which was the other option I considered

File: src/exchange.rs [hunk 7]
```diff
@@ -106,18 +121,52 @@ mod tests {
         let eur = test::find("EUR").unwrap();
         let gbp = test::find("GBP").unwrap();
 
-        let eur_usd_rate = ExchangeRate::new(usd, eur, dec!(1.5)).unwrap();
-        let eur_gbp_rate = ExchangeRate::new(usd, gbp, dec!(1.6)).unwrap();
+        let usd_eur_rate = ExchangeRate::new(usd, eur, dec!(1.5)).unwrap();
+        let usd_gbp_rate = ExchangeRate::new(usd, gbp, dec!(1.6)).unwrap();
```
Comment (RestitutorOrbis) [inline src/exchange.rs:125]: The variable names were a bit confusing here

File: src/exchange.rs [hunk 8]
```diff
@@ -106,18 +121,52 @@ mod tests {
...
 
         let fetched_rate = exchange.get_rate(usd, gbp).unwrap();
         assert_eq!(fetched_rate.rate, dec!(1.6));
+
+        let fetched_inverse_rate = exchange.get_rate(gbp, usd).unwrap();
+        assert_eq!(fetched_inverse_rate.rate, dec!(1) / dec!(1.6));
+    }
+
+    #[test]
+    fn exchange_fails_to_store_inverse_of_zero_rate(){
```
Comment (RestitutorOrbis) [inline src/exchange.rs:144]: Add test to make sure storing inverse of zero rate fails

File: src/exchange.rs [hunk 9]
```diff
@@ -106,18 +121,52 @@ mod tests {
...
+
+        let mut exchange = Exchange::new();
+        let error = exchange.set_rate_and_inverse(&eur_usd_rate);
+
+        let expected = Err(MoneyError::DivideByZero);
+        assert_eq!(error, expected);
+    }
+
+    #[test]
+    fn exchange_stores_zero_rate(){
```
Comment (RestitutorOrbis) [inline src/exchange.rs:164]: Add test to ensure storing zero rate normally works

**Comment by RestitutorOrbis** [comment]
@varunsrin Are you still interesting in having this PR merged in? If not, I can close it.
